### ISSUE TITLE
Inquiry about memory usage compare to maskrcnn_benchmark repository.

### ISSUE THREAD

**Issue by jtlee90** [pytorch/vision#1150]
I cross checked that FasterRCNN in torchvision and in https://github.com/facebookresearch/maskrcnn-benchmark/ are almost identical.

I tried both repository and compared memory usage. What I found is that maskrcnn_benchmark repository use less GPU. I am using Titan xp pascal 1 GPU, and I could run it with 5 batch images for maskrcnn repo, but for torchvision, 3 batch was maximum.
Can I know why?

**Comment by fmassa** [comment]
Hi,

Indeed, the implementation of object detection started off as a fork of `maskrcnn-benchmark`.

One possibility that makes that your model uses more memory in torchvision is that we have a naive implementation of `batched_nms`, which creates a much bigger matrix for the IoU, see https://github.com/pytorch/vision/blob/fecd138504354010133f4503027d2506df8aa967/torchvision/ops/boxes.py#L30-L61

I believe that if you replace the implementation in torchvision with a simple for-loop based implementation, the memory usage should go down.

Let me know if this fixes the differences for you. If this becomes a problem I can look into further optimizing the batched nms for larger batches

**Comment by jtlee90** [comment]
Thank you for notice! I will try it. I will check time and memory usage for that and let you know.

**Comment by fmassa** [comment]
Also, there is one more thing that is worth noticing: we use a different flavour of ResNet in torchvision, where the stride goes in the 3x3 convolution, instead of in the 1x1 convolution.
This makes memory usage for ResNet slightly larger as well.

**Comment by jtlee90** [comment]
Hello again. I was on looooong vacation so that I couldn't focus on this much.
Anyway, I analyze the w/ batched_nms and w/o. It turns out that it use more memory but a little.
On 2 batch size, less than 500mb... so... I don't think this is a main problem.

I pull out all preprocess step from a model to be same as `maskrcnn-benchmark` except data_loader.  I removed aspect_ratio group loading part only and still usage differences are a lot(If aspect_ratio_group effects a lot, please let me know)

Even for batch 1 and batch 2, I observed that model use 2.5Gb and 6.6Gb respectively, which is not even a linear increasing! (I check memory usage using `torch.cuda.max_memory_allocated()`. Maybe this is because problem of image resize to deal as batch?

I also wonder that `torch.cuda.max_memory_allocated()` says always less memory than actually using it. When function said 8GB, `nvidia-smi` said 11GB, which make me confuse...

If network is almost identical, what should I have to look?

**Comment by jtlee90** [comment]
More detailed information about what I am using configuration
1. For `torchvision`, same as  torchvision detection tutorial(colab)[https://colab.research.google.com/github/pytorch/vision/blob/temp-tutorial/tutorials/torchvision_finetuning_instance_segmentation.ipynb]
2. For `maskrcnn-benchmark`, I am using this config file: https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/configs/e2e_faster_rcnn_R_50_FPN_1x.yaml

If you set batch size as 5, then you can easily reproduce this memory usage difference.
you can not run upper batch size 4 on torchvision tutorial. Hope this help you to advice me.

**Comment by jtlee90** [comment]
@fmassa Finally, I figured out what is the problem.
First, It is not a problem of the optimization. I compared # of parameters in FRCNN in torchvision and maskedrcnn-benchmark and turns out that difference is less than 1%.
So, I conclude that usage difference comes from the computation!

Finally, I found that the configuration that I was used actually employed fixed batch normalization so that they do not need to compute running means and vars in backbone.

I brought resnet and fixed the batch norm layers as freeze and then could run with batch size 4~5.
I believe that if batch norm trained with a lot of dataset, then may do not need to learn anymore.

So, if there is option to freeze more layers(torchvision.detection also freeze first layer), those models will be more easy to deploy.

Please let me know if there is misunderstanding.

**Comment by jtlee90** [comment]
@fmassa I am still waiting your response
